src/Microsoft.ML.SearchSpace/Tuner/RandomTuner.cs

```diff
@@ -11,9 +11,16 @@ internal sealed class RandomTuner
     {
         private readonly Random _rnd;
 
-        public RandomTuner()
+        public RandomTuner(int? seed = null)
```

Comment:
LDRandom (Low-discrepancy random) from TLC does well; better than standard random. 

You may want to pull it into this repo. It helps cover the search space more quickly by choosing points in the hyperparameter space further away from already tried positions.  

Random selection tends to form clumps of tried positions, which is inefficient; it oversamples in some areas while leaving large holes of unexplored space.

One of the LDRandom sampling algos: https://en.wikipedia.org/wiki/Sobol_sequence

Reply:
Thanks for the advice

@ericstj Is there any chance that we can steal the implementation of LDRandom from old TLC?

If not I can try to introduce LDRandom tuner when I have time, or @justinormont you are more than welcome to create a PR for LDRandom if you're interested.

Reply:
One of the things the team tried to do when open-sourcing ML.NET was remove redundant components as much as possible.  When it had things that were done differently because they couldn't change .NET we tried to capture those differences and raise them as features for the .NET product.  I thought Random was one  of those, but I could be wrong.  If this feature request was missed we should first raise that in dotnet/runtime.   If we think it's really important we could bring back some functionality internally in ML.NET.  cc @eerhardt @tarekgh @tannergooding

Reply:
Had a quick look at this and it was using Intel MKL, specifically [vslNewStream (intel.com)](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/statistical-functions/random-number-generators/service-routines/vslnewstream.html) and [BRNG Parameter Definition (intel.com)](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/statistical-functions/random-number-generators/basic-generators/brng-parameter-definition.html#brng-parameter-definition_TBL10-2) with either Sobol or Niedermann.  I think the ask to .NET here might be around providing more options for random number generation algorithms.

Reply:
See https://github.com/dotnet/machinelearning/pull/1786 for the work that was done to unify with .NET. The idea being that instead of ML.NET's own `IRandom` interface, we use the unsealed `System.Random` class, and provide implementations as necessary. That way code that needs a random generator can use _any_ random generator.

Here's an implementation that still exists in ML.NET:

https://github.com/dotnet/machinelearning/blob/50e50689075c0726da3726f9ee06cd575262cb7a/src/Microsoft.ML.Core/Utilities/Random.cs#L82-L86

I assume LDRandom would follow this same design/pattern.

Reply:
Related to what @ericstj says, LDRand was held back in the internal TLC repo due to a dependence on the MKL library. If I recall, MKL was removed long ago, which should make it simple to move this code.

In this case, the ask is provide stronger methods for sampling the search space than generating a N-dimensional random position. 

Here's Optuna's docs on using Sobol for a low-discrepancy sequence of hyperparameter values:
https://optuna.readthedocs.io/en/latest/reference/samplers/generated/optuna.samplers.QMCSampler.html

I suggest LDRand since it's useful (compared to grid or random) and should be easy to pull into this repo from TLC.

Reply:
> If I recall, MKL was removed long ago, which should make it simple to move this code.

I couldn't find a version of LDRand that didn't depend on MKL.  I checked the latest codebase (2019) and the fork prior OSS.  Both still used MKL.  I also checked all of the repos I can see on MSData and couldn't find any hits.  Might be some other place but I couldn't find it.

Reply:
Is there any problem with putting the LDRand implementation in https://github.com/dotnet/machinelearning/tree/main/src/Microsoft.ML.Mkl.Components?

And then here in `RandomTuner`'s constructor taking a `System.Random` instance?

That way someone external that already depends on Mkl.Components (for example Model Builder) could pass in a LDRand into this class.